### PR TITLE
[Validator] more strict e-mail validation regex

### PR DESCRIPTION
**Pull request by xabbuh**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #14848 |
| License | MIT |
| Doc PR |  |

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
:+1:
Btw, I'm wondering if we ever considered using the W3's definition of an email as specified in HTML5?
This is much simpler than the full RFC, and still is already enforced by browsers for those using the "email" input type: `/^[a-zA-Z0-9.!#$%&’*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/`

**Comment by Tobion** [comment]
You sure this is according to the standard? From what I read, the domain may be bracketed.
RFC 2822, section 3.4.1
> domain          =       dot-atom / domain-literal / obs-domain
domain-literal  =       [CFWS] "[" *([FWS] dcontent) [FWS] "]" [CFWS]

**Comment by nicolas-grekas** [comment]
See http://www.w3.org/TR/html-markup/input.email.html#input.email.attrs.value.single

File: src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php [hunk 1]
```diff
@@ -91,6 +91,7 @@ public function getInvalidEmails()
             array('example'),
             array('example@'),
             array('example@localhost'),
+            array('foo@example.com bar'),
```
Comment (iltar) [inline src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php]: http://en.wikipedia.org/wiki/Email_address

`admin@mailserver1 (local domain name with no TLD)`

I'm not sure about the RFC, but it seems that the simple test should allow the localhost example.

File: src/Symfony/Component/Validator/Constraints/EmailValidator.php [hunk 2]
```diff
@@ -24,7 +24,7 @@
 class EmailValidator extends ConstraintValidator
 {
     /**
-     * isStrict
+     * isStrict.
```
Comment (jakzal) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: I'd rather remove this comment ;)

**Comment by xabbuh** [comment]
>  Btw, I'm wondering if we ever considered using the W3's definition of an email as specified in HTML5?

Why not? But that should imo be considered a new feature and therefore be introduced in Symfony 2.8.

File: src/Symfony/Component/Validator/Constraints/EmailValidator.php (hunk 2 above)
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: fair enough :)

**Comment by xabbuh** [comment]
Actually, @iltar is right about localhost being a valid domain-part. The HTML5 pattern would accept that. Does everybody agree with switching to that regular expression then?

**Comment by fabpot** [comment]
We had a very lengthy discussion about this a few months ago, changing anything in the current regex would be a BC break.

**Comment by xabbuh** [comment]
@fabpot Do you mean even the change forbidding whitespaces in the domain part would be a BC break?

**Comment by fabpot** [comment]
Not necessarily. I just wanted to say that it would be very easy to break BC if we change the current regex we agree upon a while ago.

**Comment by nicolas-grekas** [comment]
I'd be -1 on making the current rx even more weak. i.e. removing the "contains dot" constraint. It has been fine for months.

**Comment by xabbuh** [comment]
I didn't mean to remove the check for the period in the domain part. I meant that the HTML5 pattern that you suggested also addresses the issue @iltar raised.

**Comment by xabbuh** [comment]
So, what do we do here now?

**Comment by nicolas-grekas** [comment]
I think this one is conservative enough to be merges asis
ping @symfony/deciders

**Comment by Tobion** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Pull request merged**